### ISSUE TITLE
Transcript Versions - Gaps and replace alignments with newer annotations

### ISSUE THREAD

**Issue by davmlaw** [SACGF/variantgrid#494]
The alignments can change for the same transcript version, so we should update data to the latest (if using --replace)

https://www.ncbi.nlm.nih.gov/nuccore/NM_004656.4 is 3600bp long, and 1 base short summing up JSON exon lengths

Emma noticed that GCF_000001405.39_GRCh38.p13_genomic.109.20210514.gff.gz the exon lengths sum to 3600 - ie different than our JSON data.

Here's the genePred data (0-based like the JSON) and yes it looks like the 1st (in genomic order) exon start is 52435023 while the JSON is 52435024

```
$ zgrep NM_004656 GCF_000001405.25_GRCh37.p13_genomic.105.20201022.genePred
NM_004656.4     NC_000003.11    -       52435023        52444024        52436303        52443894        17      52435023,52436617,52436794,52437153,52437431,52438468,52439125,52439780,52440268,52440844,52441189,52441414,52441973,52442489,52443569,52443729,52443857,  52436437,52436690,52436887,52437314,52437910,52438602,52439310,52439928,52440392,52440923,52441332,52441476,52442093,52442622,52443624,52443759,52444024,  0       BAP1    cmpl    cmpl    1,0,0,1,2,0,1,0,2,1,2,0,0,2,1,1,0,
```

The data was originally inserted via these files which do have the 52435024 coordinate:

```
ucsc_hg19.20181211.genePred:NM_004656.4 chr3    -       52435024        52444024        52436303        52443894        17      52435024,52436617,52436794,52437153,52437431,52438468,52439125,52439780,52440268,52440844,52441189,52441414,52441973,52442489,52443569,52443729,52443857,  52436437,52436690,52436887,52437314,52437910,52438602,52439310,52439928,52440392,52440923,52441332,52441476,52442093,52442622,52443624,52443759,52444024,  0       BAP1
ucsc_hg19.20190821.genePred:NM_004656.4 chr3    -       52435024        52444024        52436303        52443894        17      52435024,52436617,52436794,52437153,52437431,52438468,52439125,52439780,52440268,52440844,52441189,52441414,52441973,52442489,52443569,52443729,52443857,  52436437,52436690,52436887,52437314,52437910,52438602,52439310,52439928,52440392,52440923,52441332,52441476,52442093,52442622,52443624,52443759,52444024,  0       BAP1
```

The current process is:

* Go through all the GTF/GFF (and genePred) files from oldest to newest, and insert new transcript version records that we don't have
* If the old record is missing data, or aligned to a patch, then replace data with the new one

By default we do NOT replace the transcript data with the new stuff, my thoughts were that if the transcript version wasn't bumped, then that wouldn't change.

Looks like this was a bad assumption as for whatever reason, the genome being patched or UCSC having bad alignments, the alignments for the same transcript version can indeed change. Will move this to a new issue

**Comment by davmlaw** [comment]
Need to remember to re-match any classifications against transcripts that change because of this issue.

**Comment by davmlaw** [comment]
I think I'll just handle the gaps in PyHGVS

* Remove alignment_gap boolean - this is too course and uninformative - we should store info in the transcript version data and decide then whether it can be used
* It looks like cDNA_match line up with exons, which makes it pretty easy actually, just need to add a separate list in parallel with exons that stores the ```Gap=M185 I3 M250``` attribute

https://github.com/The-Sequence-Ontology/Specifications/blob/master/gff3.md

Then just need to handle that shift in the exons - probably just write that into PyHGVS - we're already made modifications etc

Previous discussion:

[3132  HGVS - mRNA/genome gapped alignments](https://github.com/SACGF/variantgrid_private/issues/3132)
[480 UTA transcripts with gaps](https://github.com/SACGF/variantgrid/issues/480)

**Comment by davmlaw** [comment]
GCF_000001405.25_GRCh37.p13_genomic.105.20201022.gff.gz.json.gz

```
import gzip
import json
from genes.models import TranscriptVersionSequenceInfo

expected_lengths = {}
for tvsi in TranscriptVersionSequenceInfo.objects.all():
    expected_lengths[tvsi.accession] = tvsi.length

filename = "/media/dlawrence/SpinningIron/reference/variantgrid_setup_data/gene_annotation/refseq/GRCh37/GCF_000001405.25_GRCh37.p13_genomic.105.20201022.gff.gz.json.gz"
with gzip.open(filename) as f:
    data = json.load(f)

num_same_length = 0
num_diff_length = 0
num_diff_with_cdna_match = 0

for transcript_accession, transcript in data["transcripts_by_id"].items():
    if expected_length := expected_lengths.get(transcript_accession):
        length = pyreference_transcript_length(transcript)
        if length == expected_length:
            num_same_length += 1
        else:
            num_diff_length += 1
            if "cDNA_match" in transcript["features_by_type"]:
                num_diff_with_cdna_match += 1
                print(f"{num_same_length=} {num_diff_length=} {num_diff_with_cdna_match=}")                                                                                                        
```

Output:

```
num_same_length=71986 num_diff_length=6016 num_diff_with_cdna_match=1073
```

**Comment by davmlaw** [comment]
I'm going to store URLs rather than filenames, existing stuff is:

```
    GFF_URLS = {
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.60.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-60/gtf/homo_sapiens/Homo_sapiens.GRCh37.60.gtf.gz",
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.65.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-65/gtf/homo_sapiens/Homo_sapiens.GRCh37.65.gtf.gz",
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.70.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-70/gtf/homo_sapiens/Homo_sapiens.GRCh37.70.gtf.gz",
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.75.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz",
        # 81 is 75 again
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.82.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/grch37/release-82/gff3/homo_sapiens/Homo_sapiens.GRCh37.82.gff3.gz",
        # 83 has no data
        # 84 is 82 again
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.85.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/grch37/release-85/gff3/homo_sapiens/Homo_sapiens.GRCh37.85.gff3.gz",
        # 86 is 85 again
        "./ensembl/GRCh37/Homo_sapiens.GRCh37.87.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/grch37/release-87/gff3/homo_sapiens/Homo_sapiens.GRCh37.87.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.100.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-100/gff3/homo_sapiens/Homo_sapiens.GRCh38.100.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.101.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-101/gff3/homo_sapiens/Homo_sapiens.GRCh38.101.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.102.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-102/gff3/homo_sapiens/Homo_sapiens.GRCh38.102.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.103.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-103/gff3/homo_sapiens/Homo_sapiens.GRCh38.103.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.104.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-104/gff3/homo_sapiens/Homo_sapiens.GRCh38.104.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.76.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-76/gtf/homo_sapiens/Homo_sapiens.GRCh38.76.gtf.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.77.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-77/gtf/homo_sapiens/Homo_sapiens.GRCh38.77.gtf.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.78.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-78/gtf/homo_sapiens/Homo_sapiens.GRCh38.78.gtf.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.79.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-79/gtf/homo_sapiens/Homo_sapiens.GRCh38.79.gtf.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.80.gtf.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-80/gtf/homo_sapiens/Homo_sapiens.GRCh38.80.gtf.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.81.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-81/gff3/homo_sapiens/Homo_sapiens.GRCh38.81.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.82.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-82/gff3/homo_sapiens/Homo_sapiens.GRCh38.82.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.83.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-83/gff3/homo_sapiens/Homo_sapiens.GRCh38.83.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.84.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-84/gff3/homo_sapiens/Homo_sapiens.GRCh38.84.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.85.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-85/gff3/homo_sapiens/Homo_sapiens.GRCh38.85.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.86.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-86/gff3/homo_sapiens/Homo_sapiens.GRCh38.86.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.87.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-87/gff3/homo_sapiens/Homo_sapiens.GRCh38.87.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.88.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-88/gff3/homo_sapiens/Homo_sapiens.GRCh38.88.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.89.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-89/gff3/homo_sapiens/Homo_sapiens.GRCh38.89.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.90.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-90/gff3/homo_sapiens/Homo_sapiens.GRCh38.90.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.91.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-91/gff3/homo_sapiens/Homo_sapiens.GRCh38.91.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.92.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-92/gff3/homo_sapiens/Homo_sapiens.GRCh38.92.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.93.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-93/gff3/homo_sapiens/Homo_sapiens.GRCh38.93.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.94.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-94/gff3/homo_sapiens/Homo_sapiens.GRCh38.94.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.95.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-95/gff3/homo_sapiens/Homo_sapiens.GRCh38.95.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.96.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-96/gff3/homo_sapiens/Homo_sapiens.GRCh38.96.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.97.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-97/gff3/homo_sapiens/Homo_sapiens.GRCh38.97.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.98.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-98/gff3/homo_sapiens/Homo_sapiens.GRCh38.98.gff3.gz",
        "./ensembl/GRCh38/Homo_sapiens.GRCh38.99.gff3.gz.json.gz": "ftp://ftp.ensembl.org/pub/release-99/gff3/homo_sapiens/Homo_sapiens.GRCh38.99.gff3.gz",
        "./refseq/GRCh37/GCF_000001405.25_GRCh37.p13_genomic.105.20190906.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/105.20190906/GCF_000001405.25_GRCh37.p13/GCF_000001405.25_GRCh37.p13_genomic.gff.gz",
        "./refseq/GRCh37/GCF_000001405.25_GRCh37.p13_genomic.105.20201022.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/105.20201022/GCF_000001405.25_GRCh37.p13/GCF_000001405.25_GRCh37.p13_genomic.gff.gz",
        "./refseq/GRCh37/ref_GRCh37.p10_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.104/GFF/ref_GRCh37.p10_top_level.gff3.gz",
        "./refseq/GRCh37/ref_GRCh37.p13_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.105/GFF/ref_GRCh37.p13_top_level.gff3.gz",
        "./refseq/GRCh37/ref_GRCh37.p5_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/BUILD.37.3/GFF/ref_GRCh37.p5_top_level.gff3.gz",
        "./refseq/GRCh37/ref_GRCh37.p9_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.103/GFF/ref_GRCh37.p9_top_level.gff3.gz",
        "./refseq/GRCh38/GCF_000001405.38_GRCh38.p12_genomic.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109/GCF_000001405.38_GRCh38.p12/GCF_000001405.38_GRCh38.p12_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20190607.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20190607/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20190905.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20190905/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20191205.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20191205/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20200228.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20200228/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20200522.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20200522/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20200815.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20200815/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20201120.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20201120/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20210226.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20210226/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/GCF_000001405.39_GRCh38.p13_genomic.109.20210514.gff.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/annotation_releases/109.20210514/GCF_000001405.39_GRCh38.p13/GCF_000001405.39_GRCh38.p13_genomic.gff.gz",
        "./refseq/GRCh38/ref_GRCh38.p12_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.109/GFF/ref_GRCh38.p12_top_level.gff3.gz",
        "./refseq/GRCh38/ref_GRCh38.p2_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.107/GFF/ref_GRCh38.p2_top_level.gff3.gz",
        "./refseq/GRCh38/ref_GRCh38.p7_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.108/GFF/ref_GRCh38.p7_top_level.gff3.gz",
        "./refseq/GRCh38/ref_GRCh38_top_level.gff3.gz.json.gz": "http://ftp.ncbi.nlm.nih.gov/genomes/archive/old_refseq/Homo_sapiens/ARCHIVE/ANNOTATION_RELEASE.106/GFF/ref_GRCh38_top_level.gff3.gz",
    }
```

**Comment by davmlaw** [comment]
working in branch "gene_import"

Write the merge code, test with something like:

```
export GA_DIR=/media/dlawrence/SpinningIron/reference/variantgrid_setup_data/gene_annotation
python3.8 manage.py import_gene_annotation2 --annotation-consortium=RefSeq --genome-build=GRCh37 --save-merged-file="vg_gene_annotation.refseq.grch37.json.gz"  --pyreference-json ${GA_DIR}/refseq/GRCh37/ref_GRCh37.p5_top_level.gff3.gz.json.gz --pyreference-json ${GA_DIR}/refseq/GRCh37/ref_GRCh37.p9_top_level.gff3.gz.json.gz --pyreference-json ${GA_DIR}/refseq/GRCh37/ref_GRCh37.p10_top_level.gff3.gz.json.gz --pyreference-json ${GA_DIR}/refseq/GRCh37/ref_GRCh37.p13_top_level.gff3.gz.json.gz --pyreference-json ${GA_DIR}/refseq/GRCh37/GCF_000001405.25_GRCh37.p13_genomic.105.20190906.gff.gz.json.gz --pyreference-json ${GA_DIR}/refseq/GRCh37/GCF_000001405.25_GRCh37.p13_genomic.105.20201022.gff.gz.json.gz
```

**Comment by davmlaw** [comment]
TODO:

Not dealing with gene/transcripts properly as one getting lost:

```
  File "/home/dlawrence/localwork/variantgrid/genes/management/commands/import_gene_annotation2.py", line 90, in _convert_to_merged_data
    "gene_version": transcript_gene_version[transcript_accession],
KeyError: 'XR_132902.1'
```

* Is NG_/XR issue now solved?
* In genes/scripts download_gene_annotation.sh etc - have the option to run ```pyreference_gff_to_json.py``` (passing in url just used) to build data as well, maybe flags to turn on/off whether to download/generate or both etc
* Work out what’s going on in lengths not matching and no partial etc - any other way to tell? 
* PyReference pip release? Also say in README about supported GFFs?
* PyReference 5PUTR - if already have five_prime_UTR as some GFFs do - then rename instead of calculate it? Also coding_start / coding_stop etc.

**Comment by davmlaw** [comment]
Checked in start of gap handling code to https://github.com/SACGF/hgvs

Using test data that worked with VEP:

2:73385942 A>T: VEP: NM_015120.4:c.74A>T - OLD pyhgvs: NM_015120.4(ALMS1):c.74A>T
2:73385943 A>T: VEP: NM_015120.4:c.78A>T - OLD pyhgvs: NM_015120.4(ALMS1):c.75A>T
2:73385944 G>C: VEP: NM_015120.4:c.79G>C - OLD pyhgvs: NM_015120.4(ALMS1):c.76G>C

```
from genes.models import GeneVersion, TranscriptVersion
from genes.hgvs import HGVSMatcher, GenomeBuild
transcript_version = TranscriptVersion.get("NM_015120.4", GenomeBuild.grch38(), 'R')
matcher = HGVSMatcher(GenomeBuild.grch38())

matcher._pyhgvs_get_variant_tuple("NM_015120.4(ALMS1):c.74A>T", transcript_version)
Out[5]: ('2', 73385942, 'A', 'T')

matcher._pyhgvs_get_variant_tuple("NM_015120.4(ALMS1):c.75A>T", transcript_version)
matcher._pyhgvs_get_variant_tuple("NM_015120.4(ALMS1):c.76A>T", transcript_version)
matcher._pyhgvs_get_variant_tuple("NM_015120.4(ALMS1):c.77A>T", transcript_version)
# These 3 die with ValueError: Coordinate inside insertion (I3) - no mapping possible!

matcher._pyhgvs_get_variant_tuple("NM_015120.4(ALMS1):c.78A>T", transcript_version)
Out[9]: ('2', 73385943, 'A', 'T')
```

Still TODO:
* Need to handle the other way - ie genomic_to_cdna_coord
* Need to test on -'ve strand, not sure I got those right way around.

**Comment by davmlaw** [comment]
Negative strand test data

FOXI3 - NM_001135649.3 (GRCh37)

This has 3 exons but only 2 cdna_match entries. The GFF3 specs say you can even have the entire alignment represented by a single cDNA_match entry, with deletions as the introns. They only split it up so that you can get different scores:


```
2       BestRefSeq      exon    88751754        88752211        .       -       .       ID=exon-NM_001135649.3-1;Parent=rna-NM_001135649.3;Dbxref=GeneID:344167,Genbank:NM_001135649.3,HGNC:HGNC:35123,MIM:612351;Note=The RefSeq transcript has 1 frameshift compared to this genomic sequence;exception=annotated by transcript or proteomic data;gbkey=mRNA;gene=FOXI3;inference=similar to RNA sequence%2C mRNA (same species):RefSeq:NM_001135649.3;product=forkhead box I3;tag=RefSeq Select;transcript_id=NM_001135649.3
2       BestRefSeq      exon    88751415        88751751        .       -       .       ID=exon-NM_001135649.3-2;Parent=rna-NM_001135649.3;Dbxref=GeneID:344167,Genbank:NM_001135649.3,HGNC:HGNC:35123,MIM:612351;Note=The RefSeq transcript has 1 frameshift compared to this genomic sequence;exception=annotated by transcript or proteomic data;gbkey=mRNA;gene=FOXI3;inference=similar to RNA sequence%2C mRNA (same species):RefSeq:NM_001135649.3;product=forkhead box I3;tag=RefSeq Select;transcript_id=NM_001135649.3
2       BestRefSeq      exon    88746306        88748348        .       -       .       ID=exon-NM_001135649.3-3;Parent=rna-NM_001135649.3;Dbxref=GeneID:344167,Genbank:NM_001135649.3,HGNC:HGNC:35123,MIM:612351;Note=The RefSeq transcript has 1 frameshift compared to this genomic sequence;exception=annotated by transcript or proteomic data;gbkey=mRNA;gene=FOXI3;inference=similar to RNA sequence%2C mRNA (same species):RefSeq:NM_001135649.3;product=forkhead box I3;tag=RefSeq Select;transcript_id=NM_001135649.3
2       RefSeq  cDNA_match      88751415        88752211        793.757 -       .       ID=3fa72568903f70f9fc65b30792d545e3;Target=NM_001135649.3 1 798 +;assembly_bases_aln=2075;assembly_bases_seq=2075;consensus_splices=2;exon_identity=0.999648;for_remapping=2;gap_count=1;identity=0.999648;idty=0.998747;matches=2840;num_ident=2840;num_mismatch=0;pct_coverage=99.9648;pct_coverage_hiqual=99.9648;pct_identity_gap=99.9648;pct_identity_ungap=100;product_coverage=1;rank=1;splices=2;weighted_identity=0.999298;Gap=M460 I1 M337
2       RefSeq  cDNA_match      88746306        88748348        2043    -       .       ID=3fa72568903f70f9fc65b30792d545e3;Target=NM_001135649.3 799 2841 +;assembly_bases_aln=2075;assembly_bases_seq=2075;consensus_splices=2;exon_identity=0.999648;for_remapping=2;gap_count=1;identity=0.999648;idty=1;matches=2840;num_ident=2840;num_mismatch=0;pct_coverage=99.9648;pct_coverage_hiqual=99.9648;pct_identity_gap=99.9648;pct_identity_ungap=100;product_coverage=1;rank=1;splices=2;weighted_identity=0.999298
```

So, TODO:

* Need to also store cdna_start/cdna_end which is the last 2 coordinates in  "Target=NM_001135649.3 1 798 +"
* In the merge code, need to keep all the cdna_stuff
* PyHGVS - separate the link between cdna_match and exons, they won't always be 1:1
* I think do the conversion in a separate code path from exons if we have cDNA_match

**Comment by davmlaw** [comment]
Insert gene/transcript annotation data:

```
python3.8 manage.py import_gene_annotation --annotation-consortium=Ensembl --genome-build=GRCh37 --merged-json ${gene_annotation_dir}/vg_gene_annotation.ensembl.grch37.json.gz
python3.8 manage.py import_gene_annotation --annotation-consortium=Ensembl --genome-build=GRCh38 --merged-json ${gene_annotation_dir}/vg_gene_annotation.ensembl.grch38.json.gz

python3.8 manage.py import_gene_annotation --annotation-consortium=RefSeq --genome-build=GRCh37 --merged-json ${gene_annotation_dir}/vg_gene_annotation.refseq.grch37.json.gz
python3.8 manage.py import_gene_annotation --annotation-consortium=RefSeq --genome-build=GRCh38 --merged-json ${gene_annotation_dir}/vg_gene_annotation.refseq.grch38.json.gz
```

**Comment by TheMadBug** [comment]
Think this has been well tested by fire effectively.
